### ISSUE TITLE
strict_thresholding=True detects with thresholds outside the limits of the dataset

### ISSUE THREAD

**Issue by w-k-jones** [tobac-project/tobac#315]
### Discussed in https://github.com/tobac-project/tobac/discussions/314

<div type='discussions-op-text'>

<sup>Originally posted by **harrietgilmour** July 27, 2023</sup>
Hi all,

I’ve started trying out the multiple n_min_threshold (#208) and strict thresholding (#283) features in the new version release (1.5.0) and I’m getting some slightly strange results so I thought I’d ask on here to see if it’s something I’m doing wrong or if it’s a strange bug. 

I’m running tobac over a month of model data over South America to track MCSs. As a result, I wanted to make use of #208 to give a different area requirement to the deep convective core within the larger cloud shield. Here, I am using threshold = [240, 200] and n_min_threshold = [1975, 10]. This seems to work well and as expected. 

However, the issues arise when I also specify ‘strict_thresholding = True’ (#283). At the time, I was also testing the sensitivity of the cold core / deep convective core threshold (200 in above example). To do this, I reduced the value by 10 for each run. What was strange is that tobac detected features and tracks below the minimum value of the whole datatset (i.e. the minimum brightness temp value of the dataset is 179.43K, yet tobac still detected features as reaching the 170K threshold. This continues to occur as I reduce the cold core threshold even lower. 

Perhaps this screenshot below explains this more clearly:
![tobac_170K_issue](https://github.com/tobac-project/tobac/assets/90762936/7755a032-62dc-4e4f-8e9f-a62ef3d65bed)

You can see here that the minimum brightness temp value over the whole month is 179.43K. With strict_thresholding=False, there are no features that are detected as reaching the 170K threshold (as expected). However, with strict_thresholding=True, there are 2707 features detected below the 170K threshold. This shouldn’t occur as this is below the minimum value of the dataset. 

Similarly with the tracks, tobac detects 44 tracks when 0 should be detected (as we get when strict_thresholding=False). 

I then wanted to see where exactly tobac was finding these <170K tracks. They certainly aren’t where I’d expect the lowest brightness temperature values to be, even if 170K was reached and is making me even more confused!

![tobac_map](https://github.com/tobac-project/tobac/assets/90762936/716b87e5-145f-4c9d-a5e7-b4a74d26241c)

I should also add that this is my features set up (definitely using target=’minimum’):
![tobac_feature_setup](https://github.com/tobac-project/tobac/assets/90762936/c337fabc-f2bd-41c0-b36b-25b8a02ffdf0)

I’d love to hear that I’m just doing something wrong so if I am, please let me know! Otherwise, it may be some strange bug in the code? I’m a bit stuck for ideas so a fresh pair of eyes on this would be really helpful.

Thanks :) 

</div>

**Comment by w-k-jones** [comment]
I'm not sure this has anything to do with the threshold value. Strict-thresholding ends up covering most of the domain with a single feature, even when the threshold is less than the max value/greater than the min value. I'll look into this a bit further

**Comment by w-k-jones** [comment]
This bug only occurs when `target="minimum"` and `strict_thresholding=True`, if target is maximum then everything works as it should

**Comment by w-k-jones** [comment]
Found the source of the error. When we apply strict thresholding, we set the data in regions where no features were detected at the previous threshold step to zero. Unfortunately for minimum thresholding this can result in these regions being preferentially selected rather than removed! Should be a quick fix

https://github.com/tobac-project/tobac/blob/d819159d08896cb4237db6dfb538a6944a879413/tobac/feature_detection.py#L1012

**Comment by JuliaKukulies** [comment]
Great, thank you for reacting so quickly @w-k-jones. Are you fixing this?

**Comment by w-k-jones** [comment]
Ah, I've fixed the original issue but now I've found a new issue with the strict thresholding not detecting higher threshold features which I need to look into...

**Comment by JuliaKukulies** [comment]
@w-k-jones do you mean that strict thresholding fails to detect some of the higher thresholds even though the previous thresholds are met and the features are present in the data? Is that with Harriets setup?  

I guess that one trade-off with the current implementation is that the feature detection stops as soon as the preceding threshold is not met (instead of still detecting higher thresholds for`target=maximum` when e.g. at least one of the preceding thresholds is met). But I think you meant something else?

**Comment by w-k-jones** [comment]
@JuliaKukulies yes, there's something else going on. I've been testing with the example notebooks. When using only a single value for `n_min_threshold` strict thresholding gives different results, which I don't think it should. In general it detects fewer features, the majority of which are at the first threshold level.

Example using the `Example_OLR_Tracking_model` notebook:

Parameters:
```
parameters_features['position_threshold']='weighted_diff'
parameters_features['sigma_threshold']=0.5
parameters_features['n_min_threshold']=4
parameters_features['target']='minimum'
parameters_features['threshold']=[250,225,200,175,150]
```
Without strict thresholding:
1585 features
Features at each threshold level:
250 322
225 333
200 268
175 365
150 297

With strict thresholding:
1174 features
Features at each threshold level:
250 1050
225 48
200 23
175 40
150 13

As far as I'm aware, strict threshold should give the same results if `n_min_threshold` is fixed, so I think there's some sort of bug here...

**Comment by w-k-jones** [comment]
Something about the way we are masking non-detected regions in the data is changing the feature detection at subsequent thresholds, but I can't work out why it's having this effect

**Comment by JuliaKukulies** [comment]
Oh you are right that we would want the same behaviour when `n_min_threshold` is the same for all thresholds. Could it make a difference that for `strict_thresholding=True`, we require that a feature lies within the previous feature? 

For `strict_thresholding = False` a feature of the threshold 150 would be counted for the 150 category when the previous thresholds for `n_min_threshold` are met, but it here it would be enough if it is just connected to the features of 250, 225, etc. If I understand it correctly, for  `strict_thresholding = True` we remove the areas outside of the detected features in `track_data` in each iteration so that the next (more extreme) threshold must lie within the region of the previously detected features.

**Comment by w-k-jones** [comment]
Yep, that's how I think it works. I have no idea however why the existing implementation was breaking things, it's likely some odd interaction within `feature_detection_threshold`. I'm testing a different implementation in https://github.com/w-k-jones/tobac/commit/ced3044d60cda2d7040190ec45fe1b0e6ede4b93 which seems to be working a lot better

**Comment by JuliaKukulies** [comment]
I think I found the bug in the existing implementation. 

Here, we only extract the indices of the first feature of the dictionary `regions_i` that contains all detected features with their respective regions for one specific threshold: 

https://github.com/tobac-project/tobac/blob/d819159d08896cb4237db6dfb538a6944a879413/tobac/feature_detection.py#L1008-L1010

Instead, we would probably like to extract the indices of **all** detected features to then mask the data with the valid regions in the next step. For example: `region_indices: np.ndarray = np.concatenate(list(regions_i.values())).ravel()` instead of the above line. 

But I also like your new implementation @w-k-jones and will review it in detail soon! I think it makes actually sense to handle strict thresholding in `remove_parents` as it is consistent with how we remove old features for multiple thresholds.

**Comment by w-k-jones** [comment]
Great catch! This reveals a weakness in a lot of our testing, which only tend to use one or two features. Perhaps this is something we should work on in future? Might make sense to add a testing function that both produces test features/segments/tracks and the expected output, so we can more easily test larger numbers of features

**Comment by JuliaKukulies** [comment]
I agree, @w-k-jones and good idea! Should we open a new issue for that? Maybe good to take that point to the next dev meeting as well.

**Comment by w-k-jones** [comment]
Have opened a new issue in #319 , and yes, let's discuss it at the next dev meeting!

**Comment by JuliaKukulies** [comment]
Excellent, thanks!

**Comment by freemansw1** [comment]
I'm closing this as we've resolved it with #316.
